services/m3dbnode/config/config.go

```diff
@@ -190,3 +203,134 @@ type HashingConfiguration struct {
 	// Murmur32 seed value.
 	Seed uint32 `yaml:"seed"`
 }
+
+// GetETCDConfig creates a new embedded etcd config from kv config.
+func GetETCDConfig(cfg Configuration) (*embed.Config, error) {
+	newKVCfg := embed.NewConfig()
+	kvCfg := cfg.EnvironmentConfig.KV.Server
+
+	dir := kvCfg.Dir
+	if dir == "" {
+		dir = path.Join(cfg.Filesystem.FilePathPrefix, defaultETCDDirSuffix)
+	}
+	newKVCfg.Dir = dir
+
+	// listen-peer-urls
+	LPUrls, err := convertToURLsWithDefault(kvCfg.LPUrls, defaultETCDListenHost+defaultETCDServerPort)
+	if err != nil {
+		return nil, err
+	}
+	newKVCfg.LPUrls = LPUrls
+
+	// listen-client-urls
+	LCUrls, err := convertToURLsWithDefault(kvCfg.LCUrls, defaultETCDListenHost+defaultETCDClientPort)
+	if err != nil {
+		return nil, err
+	}
+	newKVCfg.LCUrls = LCUrls
+
+	host, err := getHostFromHostID(kvCfg.InitialCluster, kvCfg.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// initial-advertise-peer-urls
+	APUrls, err := convertToURLsWithDefault(kvCfg.APUrls, host+defaultETCDServerPort)
+	if err != nil {
+		return nil, err
+	}
+	newKVCfg.APUrls = APUrls
+
+	// advertise-client-urls
+	ACUrls, err := convertToURLsWithDefault(kvCfg.ACUrls, host+defaultETCDClientPort)
+	if err != nil {
+		return nil, err
+	}
+	newKVCfg.ACUrls = ACUrls
+
+	newKVCfg.Name = kvCfg.Name
+	newKVCfg.InitialCluster = kvCfg.InitialCluster
+
+	return newKVCfg, nil
+}
+
+func convertToURLsWithDefault(rawURLs []string, def ...string) ([]url.URL, error) {
+	if len(rawURLs) == 0 {
+		rawURLs = def
+	}
+
+	var urls []url.URL
+
+	for _, u := range rawURLs {
+		parsed, err := url.Parse(u)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse URL: %s", u)
+		}
+		urls = append(urls, *parsed)
+	}
+	return urls, nil
+}
+
+func getHostFromHostID(initialCluster, hostID string) (string, error) {
+	if len(initialCluster) == 0 {
+		return "", errors.New("zero nodes in initialCluster")
+	}
+
+	nodesStr := strings.Split(initialCluster, ",")
+
+	for _, nodeStr := range nodesStr {
+		nodeData := strings.Split(nodeStr, "=")
+
+		if len(nodeData) != 2 {
+			return "", errors.New("invalid initialCluster format")
+		}
+
+		if hostID == nodeData[0] {
+			endpoint := nodeData[1]
+
+			colonIdx := strings.LastIndex(endpoint, ":")
+			if colonIdx == -1 {
+				return "", errors.New("invalid initialCluster format")
+			}
+
+			return endpoint[:colonIdx], nil
+		}
+	}
+
+	return "", errors.New("host not in initialCluster list")
+}
+
+// InitialClusterToETCDEndpoints converts the etcd config initialCluster into etcdClusters config
+// for m3cluster.
+func InitialClusterToETCDEndpoints(initialCluster string) ([]string, error) {
```

Comment:
does this need to be an exported type? any reason to not accept initialCluster as `initialClusterHosts []string` if you're having to do this anyway?

Reply:
I thought about this too, but decided to keep the same format as the `initial-cluster` config from etcd, thinking that it'll make it easier for users to construct the config.

Reply:
See my comment in `getHostFromHostID(...)`.  Unfortunately the hostID we resolve is not always going to match the raw etcd ip:port or host:port that we give to etcd when we configured it.